### DISCUSSION TITLE
Is there a way to put a song to the 1st place using API?

### DISCUSSION THREAD

**Discussion by Suoslex** [navidrome/navidrome#3258] [Q&A] [answered]
So, I tried to examine docs and google thoroughly, but didn't find anything unfortunately.
My goal is to add a song to the beginning of a playlist using API.
The only thing which covers API in the documentation is Subsonic API, and the only apparent way to add tracks is by using **/updatePlaylist** endpoint. 
The problem is that it only has **songIdToAdd** and **songIndexToRemove** keys, and when I used songIdToAdd it just adds a song to the end of playlist, which is not what I want.
So the only way I've come to is by removing all songs from the playlist by using **songIndexToRemove**, and then readding them with the new one at the first place. 
The problem is that there is a maximum length of GET request, and I need to split data to a few requests to make it work.
And I would use this way, But I feel like it's too complex. There should be an easier way (like with Navidrome API, where you send "insert_before" for the track).

Any hints will be appreciated.

P.s. It would be great if we could add tracks to the beginning of the playlist using UI as well. Because dragging it from the bottom to the beginning can really take a while sometimes.
P.p.s. I know we can reorder playlist in UI using table ordering, but in other Subsonic Clients it's not always possible, and I want to have the newest tracks always on the top.

**Comment by deluan** [comment, marked as answer]
There is currently no way to add a song to the 1st place of a playlist. But based on your ideas, here's a couple of suggestions:

1. Navidrome supports OpenSubsonic [`formPost`](https://opensubsonic.netlify.app/docs/extensions/formpost/) extension, which allow requests to be made as POST. This helps with increasing the number of IDs being sent to the playlists endpoints
2. You could create a [Smart Playlist](https://github.com/navidrome/navidrome/issues/1417), that gets songs from your regular playlist but sort it by `dateAdded` descendent. This would cause the smart playlist to be exact like the one you add songs, but in reverse.
3. Take a look at how [Feishin](https://github.com/jeffvli/feishin) handles playlists (using the Network panel in your browser inspector), I think it uses the Native API, not Subsonic's. But this would tie you to the undocumented Native API, and my changes could break your solution.

**Reply by Suoslex** [reply]
1) Is it already installed by default? Or I should install it myself?
2) Smart playlists are good, but I don't like that I can't change order of items in it. Yes, having items sorted by dateAdded is the closest thing to what I need, but sometimes even with that I want to change a position of one or another song, and it's not possible to do without some extra server-side work.
3) Yeah, it's a possible way. But I'm more inclined to use OpenSubsonic formPost extension.

Thank you for your answers, I really appreciate it.

**Reply by deluan** [reply]
Yes, Navidrome has supported POST API calls since the beginning: https://github.com/navidrome/navidrome/commit/1278863416746e6f8da0597dd7d2abeb16984c87

Also, do check how Feishin does it, as it may or may not use the Subsonic API for playlists, but FWIW it allows reordering of playlists and it is really fast when doing it.

**Reply by Suoslex** [reply]
It uses default Navidrome API. And uses the same method sending "insert_before". 

<img width="246" alt="Screenshot 2024-09-14 at 23 39 49" src="https://github.com/user-attachments/assets/fa723597-1d3e-4f58-9c17-9cd9f1747cff">
